src/file_pipe_log/pipe_builder.rs

```diff
@@ -99,108 +101,203 @@ impl<F: FileSystem> DualPipesBuilder<F> {
+                let path = file_id.build_file_path(&dir);
+                if file_system.exists_metadata(&path) {
+                    delete_start = Some(i.saturating_sub(1) * min_id / max_sample + 1);
+                    break;
+                }
+            }
+            // Delete metadata starting from the oldest. Abort on error.
+            if let Some(start) = delete_start {
+                let mut success = 0;
+                for seq in start..min_id {
+                    let file_id = FileId { queue, seq };
+                    let path = file_id.build_file_path(&dir);
+                    if file_system.exists_metadata(&path) {
+                        if let Err(e) = file_system.delete_metadata(&path) {
+                            error!("failed to delete metadata of {}: {}.", path.display(), e);
+                            break;
                         }
-                        _ => {}
+                    } else {
+                        continue;
                     }
+                    success += 1;
                 }
+                warn!(
+                    "deleted {} stale files of {:?} in range [{}, {}).",
+                    success, queue, start, min_id,
+                );
+            }
+        }
+        // Steps in the procedure of `scan`: `validate_dir` -> `parse_file_range` ->
+        // `clean_stale_metadata`
+        // Validate the main dir.
+        let dir = &self.cfg.dir;
+        validate_dir(dir)?;
+        self.dir_lock = Some(lock_dir(dir)?);
+        // Validate the secondary dir.
+        if let Some(secondary_dir) = self.cfg.secondary_dir.as_ref() {
+            validate_dir(secondary_dir)?;
+        }
+
+        type FileSeqRange = (u64, u64); /* (minimal_id, maximal_id) */
+        type FileSeqDict = HashMap<FileSeq, StorageDirType>; /* HashMap<seq, dir_type> */
+        let mut append_id_range = (u64::MAX, 0_u64);
+        let mut rewrite_id_range = (u64::MAX, 0_u64);
+        let mut append_file_dict: HashMap<FileSeq, StorageDirType> = HashMap::default();
+        let mut rewrite_file_dict: HashMap<FileSeq, StorageDirType> = HashMap::default();
+        // Parse files in `cfg.dir`.
+        parse_file_range(
+            dir,
+            StorageDirType::Main,
+            &mut append_id_range,
+            &mut rewrite_id_range,
+            &mut append_file_dict,
+            &mut rewrite_file_dict,
+        )?;
+        // Parse files in `cfg.secondary_dir`.
+        if_chain::if_chain! {
+            if let Some(secondary_dir) = self.cfg.secondary_dir.as_ref();
+            if let Ok(0) = parse_file_range(
+                    secondary_dir,
+                    StorageDirType::Secondary,
+                    &mut append_id_range,
+                    &mut rewrite_id_range,
+                    &mut append_file_dict,
+                    &mut rewrite_file_dict,
+                );
+            if let Ok(true) = from_same_dev(&self.cfg.dir, secondary_dir);
```

Comment:
This is wrong, user would've migrated the directory from another disk. Let's check the device during config validation instead.

Reply:
But I prefer doing dir checking here.

Here, only when "the `secondary-dir` had no valid log files" and "`from_same_dev` returned `true`", we could ignore the `secondary-dir`. And whether the migrated directory owns valid files has been checked by `if let Ok(0) = parse_file_range(...)`.

If we made dir checking during the config validation, we would not know _whether there exists valid log files in the `secondary-dir` when `from_same_dev(...)` returns `true`_. In this implementation,  we should check it again when scanning the two dir and make the decision whether we could ignore the `secondary-dir`.

Reply:
So what's the point of doing the check at all? We already allow a secondary directory on the same device in the case of "user migration". The way I see it, you either deny it or accept it, there shouldn't be any middle ground.

Reply:
No, we don't allow user use a `secondary-dir` on the same device, the strategy here is to check and ignore this case, by `reset cfg.secondary-dir = None` and logging with warning `"sub-dir ({}) and dir ({}) are on same device, ignore it"`.

Maybe I should add an extra specification to `secondary-dir`, which is only valid when it's on different device?

Reply:
> Here, only when "the secondary-dir had no valid log files" and "from_same_dev returned true", we could ignore the secondary-dir.

You said it before, this case is tolerated, i.e. the secondary directory on the same device can still be used.

Reply:
Emmm.....more precisely, I expect and strongly recommend that users should use different dir on different devices respectively.

However, as u mentioned above, some users might migrate the `secondary-dir` from another device to the device, `dir` belonging to. And in this case, I think we should tolerate it, considering data integrity. And that's why I introduce this check here.

Reply:
My point is, if the code can still work (and is expected to work) if the secondary-dir is located on the same device, then there's no need to impose a restriction. By banning this case, you are in fact making it more difficult to test. That's why I said "you either deny it or accept it, there shouldn't be any middle ground".

Of course, you can still emit a warning (preferably during config validation, you can call `read_dir` there to detect if the directory is empty).